mcs/class/corlib/System.Threading/Interlocked.cs

```diff
@@ -92,9 +122,28 @@ public static partial class Interlocked
 		public extern static double CompareExchange(ref double location1, double value, double comparand);
 
 		[ComVisible (false)]
+		[ReliabilityContract (Consistency.WillNotCorruptState, Cer.Success)]
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
+		[return: NotNullIfNotNull("location1")]
+		extern static void CompareExchange_T (ref T location1, ref T value, ref T comparand, ref T result) where T : class?;
+
+		[ComVisible (false)]
 		[ReliabilityContract (Consistency.WillNotCorruptState, Cer.Success)]
-		public extern static T CompareExchange<T> (ref T location1, T value, T comparand) where T:class;
+		[return: NotNullIfNotNull("location1")]
+		public static T CompareExchange<T> (ref T location1, T value, T comparand) where T : class?
+		{
```

Comment:
Without the ability to make locals volatile, this seems fragile under heavy optimization.